pkg/remote/message.go

```diff
@@ -172,7 +178,32 @@ func (m *message) RPCInfo() rpcinfo.RPCInfo {
 
 // ServiceInfo implements the Message interface.
 func (m *message) ServiceInfo() *serviceinfo.ServiceInfo {
-	return m.svcInfo
+	return m.targetSvcInfo
+}
+
+func (m *message) SpecifyServiceInfo(svcName, methodName string) (*serviceinfo.ServiceInfo, error) {
+	// for non-multi-service including generic server scenario
+	if m.targetSvcInfo != nil {
+		if mt := m.targetSvcInfo.MethodInfo(methodName); mt == nil {
+			return nil, NewTransErrorWithMsg(UnknownMethod, fmt.Sprintf("unknown method %s", methodName))
+		}
+		return m.targetSvcInfo, nil
+	}
+	if svcName == "" && m.refuseTrafficWithoutServiceName {
+		return nil, NewTransErrorWithMsg(NoServiceName, "no service name while the server has WithRefuseTrafficWithoutServiceName option enabled")
+	}
+	var key string
+	if svcName == "" {
+		key = methodName
+	} else {
+		key = fmt.Sprintf("%s.%s", svcName, methodName)
```

Comment:
Also for `fmt.Sprintf` here (and the rest occurrences), please add a helper func for this with `strings.StringBuilder`.